README.md
# Color::FFI

example color class written using Module::Build::FFI

# SYNOPSIS

    my $color = Color::FFI->new(255,255,255); # same as Color::FFI->white;
    $color->set_green(75);
    $color->set_blue(100);
    say "red is ", $color->get_red;
    say "green is ", $color->get_green;
    say "blue is ", $color->get_blue;

# DESCRIPTION

This module is intended as an example to allow me to experiment with 
[Module::Build::FFI](https://metacpan.org/pod/Module::Build::FFI), [FFI::Platypus](https://metacpan.org/pod/FFI::Platypus).  It isn't intended to be 
useful.

Objects of this class represent a color stored as three 8-bit color 
channel values, red, green and blue.

# CONSTRUCTOR

## new

To create an instance of this class, you can use new, or one of the 
convenience methods.  If you use new you must provide an integer triplet 
of the red, green and blue values.

    my $color = Color::FFI->new(0x7f,0xff,0x00);
    my $red = Color::FFI->red;
    my $green = Color::FFI->green;
    my $blue = Color::FFI->blue;

# METHODS

## white

    my $color = Color::FFI->white;

Class method that returns an instance of [Color::FFI](https://metacpan.org/pod/Color::FFI) that is pure
white.

## black

    my $color = Color::FFI->black;

Class method that returns an instance of [Color::FFI](https://metacpan.org/pod/Color::FFI) that is pure
black.

## red

    my $color = Color::FFI->red;

Class method that returns an instance of [Color::FFI](https://metacpan.org/pod/Color::FFI) that is pure
red.

## get\_red

    my $red = $color->get_red;

Return the red color channel as an 8-bit value.

## set\_red

    $color->set_red($red);

Sets the red color channel as an 8-bit value.

## green

    my $color = Color::FFI->green;

Class method that returns an instance of [Color::FFI](https://metacpan.org/pod/Color::FFI) that is pure
green.

## get\_green

    my $green = $color->get_green;

Return the green color channel as an 8-bit value.

## set\_green

    $color->set_green($green);

Sets the green color channel as an 8-bit value.

## blue

    my $color = Color::FFI->blue;

Class method that returns an instance of [Color::FFI](https://metacpan.org/pod/Color::FFI) that is pure
blue.

## get\_blue

    my $blue = $color->get_blue;

Return the blue color channel as an 8-bit value.

## set\_blue

    $color->set_blue($blue);

Sets the blue color channel as an 8-bit value.

# CAVEATS

It's just an example for testing, it isn't really useful for
anything.

# AUTHOR

Graham Ollis <plicease@cpan.org>

# COPYRIGHT AND LICENSE

This software is copyright (c) 2013 by Graham Ollis.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.


Changes
Revision history for {{$dist->name}}

{{$NEXT}}

0.02      2015-02-02 06:37:25 -0500
  - remove use of inlined functions
  - use ExtUtils::CppGuess for C++ compiler / link options at install time

0.01      2015-02-02 05:42:19 -0500
  - initial version


ffi/color.cpp
class Color {

public:
  Color(int, int, int);
  ~Color();

  int get_red();
  int get_green();
  int get_blue();

  void set_red(int);
  void set_green(int);
  void set_blue(int);

  static Color *red();
  static Color *green();
  static Color *blue();
  static Color *white();
  static Color *black();

  static int _sizeof();

private:
  int _red, _green, _blue;
};

Color::Color(int r, int g, int b) : _red(r), _green(g), _blue(b)
{
}

Color::~Color()
{
}

void Color::set_red(int r)
{
  _red = r;
}

void Color::set_green(int g)
{
  _green = g;
}

void Color::set_blue(int b)
{
  _blue = b;
}

int Color::get_red()   { return _red; }
int Color::get_green() { return _green; }
int Color::get_blue()  { return _blue; }

Color *Color::red()   { return new Color(255,0,0); }
Color *Color::green() { return new Color(0,255,0); }
Color *Color::blue()  { return new Color(0,0,255); }
Color *Color::white() { return new Color(0xff,0xff,0xff); }
Color *Color::black() { return new Color(0x00,0x00,0x00); }

int Color::_sizeof() { return sizeof(Color); }


lib/Color/FFI.pm
package Color::FFI;

use strict;
use warnings;
use FFI::Platypus;
use FFI::Platypus::Memory;

# ABSTRACT: example color class written using Module::Build::FFI
# VERSION

=head1 SYNOPSIS

 my $color = Color::FFI->new(255,255,255); # same as Color::FFI->white;
 $color->set_green(75);
 $color->set_blue(100);
 say "red is ", $color->get_red;
 say "green is ", $color->get_green;
 say "blue is ", $color->get_blue;

=head1 DESCRIPTION

This module is intended as an example to allow me to experiment with 
L<Module::Build::FFI>, L<FFI::Platypus>.  It isn't intended to be 
useful.

Objects of this class represent a color stored as three 8-bit color 
channel values, red, green and blue.

=head1 CONSTRUCTOR

=head2 new

To create an instance of this class, you can use new, or one of the 
convenience methods.  If you use new you must provide an integer triplet 
of the red, green and blue values.

 my $color = Color::FFI->new(0x7f,0xff,0x00);
 my $red = Color::FFI->red;
 my $green = Color::FFI->green;
 my $blue = Color::FFI->blue;

=head1 METHODS

=head2 white

 my $color = Color::FFI->white;

Class method that returns an instance of L<Color::FFI> that is pure
white.

=head2 black

 my $color = Color::FFI->black;

Class method that returns an instance of L<Color::FFI> that is pure
black.

=head2 red

 my $color = Color::FFI->red;

Class method that returns an instance of L<Color::FFI> that is pure
red.

=head2 get_red

 my $red = $color->get_red;

Return the red color channel as an 8-bit value.

=head2 set_red

 $color->set_red($red);

Sets the red color channel as an 8-bit value.

=head2 green

 my $color = Color::FFI->green;

Class method that returns an instance of L<Color::FFI> that is pure
green.

=head2 get_green

 my $green = $color->get_green;

Return the green color channel as an 8-bit value.

=head2 set_green

 $color->set_green($green);

Sets the green color channel as an 8-bit value.

=head2 blue

 my $color = Color::FFI->blue;

Class method that returns an instance of L<Color::FFI> that is pure
blue.

=head2 get_blue

 my $blue = $color->get_blue;

Return the blue color channel as an 8-bit value.

=head2 set_blue

 $color->set_blue($blue);

Sets the blue color channel as an 8-bit value.

=head1 CAVEATS

It's just an example for testing, it isn't really useful for
anything.

=cut

my $ffi = FFI::Platypus->new;
$ffi->lang('CPP');
$ffi->package;
$ffi->custom_type( Color => {
  native_type => 'opaque',
  perl_to_native => sub { ${ $_[0] } },
  native_to_perl => sub { bless \$_[0], "Color::FFI" },
});

$ffi->attach( [ 'Color::Color(int, int, int)' => '_new' ] => ['Color', 'int', 'int', 'int'] => 'void' );
$ffi->attach( [ 'Color::~Color()' => '_DESTROY'         ] => ['Color']                      => 'void' );

sub _ffi_record_size {
  my $size = $ffi
    ->function( 'Color::_sizeof()' => [] => 'int' )
    ->call;
  die "unable to determine size of Color class"
    unless $size;
  $size;
}

sub new
{
  my $class = shift;
  my $ptr = FFI::Platypus::Memory::malloc(_ffi_record_size());
  my $self = bless \$ptr, $class;
  _new($self, @_);
  $self;
}

foreach my $color (qw( red green blue black white ))
{
  $ffi->attach( [ "Color::${color}()"        => $color        ] => []              => 'Color'  );
  next if $color =~ /^(white|black)$/;
  $ffi->attach( [ "Color::get_${color}()"    => "get_$color"  ] => ['Color']       => 'int'    );
  $ffi->attach( [ "Color::set_${color}(int)" => "_set_$color" ] => ['Color','int'] => 'void'    );
  no strict 'refs';
  *{"set_$color"} = eval qq{ sub { _set_${color}(\@_);\$_[0] } };
  die $@ if $@;
}

sub DESTROY
{
  my($self) = @_;
  _DESTROY($self);
  FFI::Platypus::Memory::free($$self);
}

1;
